docs/resource-sharing/os-backfill-containers.md

```diff
@@ -27,34 +27,145 @@ In order to configure the container, you will need:
+-   `singularity.opensciencegrid.org`
+
+cvmfsexec has the following system requirements:
+
+-   On EL7, you must have kernel version >= 3.10.0-1127 (run `uname -vr` to check), and user namespaces enabled.
+    See step 1 in the
+    [Singularity Install document](https://opensciencegrid.org/docs/worker-node/install-singularity/#enabling-unprivileged-singularity)
+    for details.
+
+-   On EL8, you must have kernel version >= 4.18 (run `uname -vr` to check).
+
+See the [cvmfsexec README](https://github.com/cvmfs/cvmfsexec#readme) details.
+
+Note that cvmfsexec will not be run if CVMFS repos are already available in `/cvmfs` via bind-mount,
+regardless of the value of `CVMFSEXEC_REPOS`.
+
+Using cvmfsexec takes place in the entrypoint, which means it will still happen
+even if you specify a different command to run, such as `bash`.
+You can bypass the entrypoint by passing `--entrypoint <cmd>` where `<cmd>` is some different command to run,
+e.g. `--entrypoint bash`.
+Setting the entrypoint this way clears the command.
+
+
+#### Additional cvmfsexec configuration
+
+There are several environment variables you can set for cvmfsexec:
+
+-   `CVMFSEXEC_REPOS` - this is a space-separated list of CVMFS repos to mount.
+    Leave this blank to disable cvmfsexec.
+    OSG jobs frequently use the OASIS repo (`oasis.opensciencegrid.org`) and
+    the singularity repo (`singularity.opensciencegrid.org`).
+
+-   `CVMFS_HTTP_PROXY` - this sets the proxy to use for CVMFS; if left blank
+    it will find the best one via WLCG Web Proxy Auto Discovery.
+
+-   `CVMFS_QUOTA_LIMIT` - the quota limit in MB for CVMFS; leave this blank to
+    use the system default (4 GB)
+
+You can add other CVMFS options by bind-mounting a config file over `/cvmfsexec/default.local`;
+note that options in environment variables take precedence over options in `/cvmfsexec/default.local`.
+
+You can store the cache outside of the container by volume mounting a directory to `/cvmfs-cache`.
+
+You can store the logs outside of the container by volume mounting a directory to `/cvmfs-logs`.
+
+
+### Adding CVMFS via bind-mount
+
+As an alternative to using cvmfsexec, you may install CVMFS on the host, and volume mount it into the container.
+This will let you avoid running the container in privileged mode.
+However, supporting Singularity jobs inside the container will require extra privileges,
+namely the capabilities `DAC_OVERRIDE`, `DAC_READ_SEARCH`, `SETGID`, `SETUID`, `SYS_ADMIN`, `SYS_CHROOT`, and `SYS_PTRACE`.
+
+Follow the [installing CVMFS document](../worker-node/install-cvmfs.md) to install CVMFS on the host.
+
+Once you have CVMFS installed and mounted on your host, add `-v /cvmfs:/cvmfs:shared` to your `docker run` invocation.
+
+This is the [example at the top of the page](#running-the-container-with-docker),
+modified to volume mount CVMFS instead of using cvmfsexec, and using reduced privileges:
+
+```
+docker run -it --rm --user osg      \
```

Comment:
```suggestion
docker run -it --rm                 \
```

`--user osg` breaks Singularity invocations within the container

Reply:
AFAIK running the container as an arbitrary non-root user was a supported use case. Does this mean we will no longer support that?

Reply:
https://github.com/opensciencegrid/osgvo-docker-pilot/pull/8#issuecomment-801495129

Reply:
Hrm, good memory so I think we have to better understand why it needs to be a non-root user in the container. I'll re-run some tests but it appears that priv Singularity fails in the `cvmfsexec` case when running as the `osg` user

Reply:
Well duh, priv Singularity invocations will fail as a non-root user: https://github.com/brianhlin/osgvo-docker-pilot/runs/3232846941?check_suite_focus=true#step:5:45

So to allow non-root usage of the container then we need https://opensciencegrid.atlassian.net/browse/SOFTWARE-4725

Reply:
Wait no, I'm not sure that makes sense. The backfill container run through Docker in privileged mode as the `osg` user with CVMFS bind mounted in works fine: https://github.com/brianhlin/osgvo-docker-pilot/runs/3232846885?check_suite_focus=true#step:5:457 but the same command with `cvmfsexec` fails with the error about not finding the setuid: https://github.com/brianhlin/osgvo-docker-pilot/runs/3232846941?check_suite_focus=true#step:5:45. OTOH, running through Docker in privileged mode as the root user with cvmfsexec works fine: https://github.com/brianhlin/osgvo-docker-pilot/runs/3225389938?check_suite_focus=true#step:5:26.

Thoughts @DrDaveD @rynge ?

Reply:
This is still using the old OSG VO wrapper, not glideinWMS, right?  That was the reason why we couldn't yet recommend that people run osgvo-docker-pilot unprivileged yet.  GlideinWMS tries the unprivileged singularity from /cvmfs/oasis.opensciencegrid.org first, I think that's the main difference for us here.

cvmfsexec itself when run as an unprivileged user uses an unprivileged user namespace, so that disables the ability to use setuid within it.  When run as root, sure, you can still run singularity inside it but then it doesn't need setuid because it is running unprivileged.  People definitely won't want to run pilot & therefore user jobs as root, however!

Reply:
Ah that makes sense. Thanks!